src/Storages/StorageMergeTree.h

```diff
@@ -151,6 +151,14 @@ class StorageMergeTree final : public MergeTreeData
     std::atomic<bool> shutdown_called {false};
     std::atomic<bool> flush_called {false};
 
+    /// PreparedSets cache for one executing mutation.
+    /// NOTE: we only store weak_ptr to PreparedSetsCache, so that the cache is shared between mutation tasks that are executed in parallel.
+    /// The goal is to avoiding consuming a lot of memory when the same big sets are used by multiple tasks at the same time.
+    /// If the tasks are executed without time overlap, we will destroy the cache to free memory, and the next task might rebuild the same sets.
+    std::mutex mutation_prepared_sets_cache_mutex;
```

Comment:
Is it possible to have several ongoing mutations? Here https://github.com/ClickHouse/ClickHouse/blob/c2611c3ba940cedf7ddc225e805e45fd77c59977/src/Storages/StorageMergeTree.h#L149 we have a map for mutations, probably the same here 🤔

Reply:
changed this to a map